### PR TITLE
[aptos-genesis] Create all accounts upfront, with initial balances

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
This allows for a file to create initial balances for any account including both in and out of the validator set.  Additionally, fixes that all accounts will have a coin store at startup.

### Test Plan
E2E tests incoming, but I've also tested this manually.  I expect I will make more correctness checks on Genesis inputs in the coming days

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3939)

### CODE REVIEW

File: aptos-move/vm-genesis/src/lib.rs [hunk 1]
```diff
@@ -82,6 +84,7 @@ pub fn encode_genesis_transaction(
     framework: &ReleaseBundle,
     chain_id: ChainId,
     genesis_config: GenesisConfiguration,
+    initial_balances: &[InitialBalance],
```
Comment (movekevin) [inline aptos-move/vm-genesis/src/lib.rs:87]: I wonder if we can add this to GenesisConfiguration? We'll have a lot more things later on like delegations, vesting, etc. so we would end up with quite a long list of args here.

File: aptos-move/vm-genesis/src/lib.rs [hunk 2]
```diff
@@ -617,10 +648,46 @@ pub fn generate_mainnet_genesis(
...
             voting_power_increase_limit: 30,
         },
+        initial_balances.as_slice(),
     );
     (genesis, test_validators)
 }
 
+/// Uses the staking amounts to determine what the correct initial balances
+/// if there are no other accounts that are initialized with coins
+pub fn parse_initial_balances_from_staking_amounts(
```
Comment (movekevin) [inline aptos-move/vm-genesis/src/lib.rs:658]: I think we should just have balances separately. The staking amount can be different from a validator's total balance (e.g. due to gas). staking amount is used specifically here https://github.com/aptos-labs/aptos-core/pull/3939/files#diff-cb728fc6a391f08ea20845355e8406cc6741c7163a388a7ad5ef0e250d1d6207R177 while initial balance is now done in the new mint step you added.

Comment (movekevin) [reply aptos-move/vm-genesis/src/lib.rs:658]: We can assertion in genesis tooling somewhere that staking amount <= initial balance

**Review by movekevin — CHANGES_REQUESTED** [review]

File: aptos-move/vm-genesis/src/lib.rs (hunk 2 above)
Comment (gregnazario) [reply aptos-move/vm-genesis/src/lib.rs:658]: We do have balances separately.

The purpose of this is so for the current testnets, testing frameworks, etc.  We then don't have to build up this file every place we do genesis, instead we build up the information from the already existing information.

Comment (gregnazario) [reply aptos-move/vm-genesis/src/lib.rs:658]: I'll replace this and weave through the separate ones later

Comment (movekevin) [reply aptos-move/vm-genesis/src/lib.rs:658]: OK sounds good 👍

**Review by movekevin — APPROVED** [review]

**Review by davidiw — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/genesis.move [hunk 3]
```diff
@@ -179,13 +190,47 @@ module aptos_framework::genesis {
...
         // validators.
         aptos_coin::destroy_mint_cap(aptos_framework);
 
         stake::on_new_epoch();
     }
 
+    /// Creates an account allowing for the signer to be used elsewhere in genesis
+    fun create_account(aptos_framework: &signer, account: address, balance: u64): signer {
+        let signer = account::create_account(account);
+        coin::register<AptosCoin>(&signer);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/genesis.move:214]: why not just use aptos_account :D

File: aptos-move/framework/aptos-framework/sources/genesis.move [hunk 4]
```diff
@@ -179,13 +190,47 @@ module aptos_framework::genesis {
...
+            aptos_coin::mint(aptos_framework, account, balance);
+        };
+        signer
+    }
+
+    /// Finds the balance of the account and removes it from the list
+    fun find_initial_balance(accounts: &mut vector<address>, balances: &mut vector<u64>, account: &address): u64 {
+        // Lookup account
+        let (found, index) = vector::index_of(accounts, account);
+        assert!(found, 404);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/genesis.move:226]: probably want to use a clearer error here, we have error::not_found(EBALANCE) or something

File: aptos-move/framework/aptos-framework/sources/genesis.move (hunk 3 above)
Comment (gregnazario) [reply aptos-move/framework/aptos-framework/sources/genesis.move:214]: It doesn't return the signer

File: aptos-move/framework/aptos-framework/sources/genesis.move (hunk 4 above)
Comment (gregnazario) [reply aptos-move/framework/aptos-framework/sources/genesis.move:226]: I'm pretty sure there isn't good error mapping in Genesis, but yes I can add that.

**Comment by davidiw** [comment]
moving in a different directions

**Pull request closed without merging**